solidity/contracts/deposit/DepositUtils.sol

```diff
@@ -458,56 +464,70 @@ library DepositUtils {
-        uint256 mainCharge = mainCharge(_d, _redeemer);
-        uint256 frtAdjust = frtAdjust(_d, _redeemer);
-        uint256 signerFeeAdjust = signerFeeAdjust(_d, _redeemer);
-        return mainCharge.add(frtAdjust).add(signerFeeAdjust);
-    }
+        bool tdtOwnerRedemption = depositOwner(_d) == _redeemer;
+        bool frtOwnerRedemption = feeRebateTokenHolder(_d) == _redeemer;
+        bool frtExists = feeRebateTokenHolder(_d) != address(0);
 
-    /// @notice             Get the base TBTC amount needed to redeem.
-    /// @param _redeemer    The deposit redeemer.
-    /// @return             The amount in TBTC.
-    function mainCharge(DepositUtils.Deposit storage _d, address _redeemer) internal view returns (uint256){
         bool inCourtesy = _d.inCourtesyCall();
-        bool ownerRedemption = depositOwner(_d) == _redeemer;
         bool atTerm = remainingTerm(_d) == 0;
+        bool redeemerPaysNoFee = inCourtesy || atTerm;
+
+        require(inCourtesy || tdtOwnerRedemption || atTerm, "Only TDT holder can redeem unless deposit is at-term or in COURTESY_CALL");
 
-        require(inCourtesy || ownerRedemption || atTerm, "Only TDT holder can redeem unless deposit is at-term or in COURTESY_CALL");
+        uint256 mainCharge = computeBaseRedemptionCharge(_d, tdtOwnerRedemption);
+        uint256 frtAdjust = computeFrtAdjustment(_d, redeemerPaysNoFee, frtOwnerRedemption, frtExists);
+        uint256 signerFeeAdjust = computeSignerFeeAdjustment(_d, redeemerPaysNoFee, frtExists);
+        return mainCharge.add(frtAdjust).add(signerFeeAdjust);
+    }
 
-        if (ownerRedemption) {
+    /// @notice                    Get the base TBTC amount needed to redeem.
+    /// @param tdtOwnerRedemption  True if the redeemer is the TDT holder.
+    /// @return                    The amount in TBTC.
+    function computeBaseRedemptionCharge(
+        DepositUtils.Deposit storage _d,
+        bool tdtOwnerRedemption
+    ) internal view returns (uint256){
+        if (tdtOwnerRedemption) {
             return 0;
         }
-
         return lotSizeTbtc(_d);
     }
 
-    /// @notice             Get the fee rebate amount needed to redeem.
-    /// @param _redeemer    The deposit redeemer.
-    /// @return             The amount in TBTC.
-    function frtAdjust(DepositUtils.Deposit storage _d, address _redeemer) internal view returns (uint256){
+    /// @notice                    Get the fee rebate amount needed to redeem.
+    /// @param _redeemerPaysNoFee  True if Deposit is at term or in courtesy call.
+    /// @param _frtOwnerRedemption True if the redeemer is the FRT holder.
+    /// @param _frtExists          True if the FRT exists.
+    /// @return                    The amount in TBTC.
+    function computeFrtAdjustment(
+        DepositUtils.Deposit storage _d,
+        bool _redeemerPaysNoFee,
+        bool _frtOwnerRedemption,
```

Comment:
Thoughts on aligning the structure here for clarity? `_redeemerPaysNoFee`, `_redeemerHoldsFrt`, `_frtExists`. I would also rather say `Owes` rather than `Pays`. This whole path is computing what is owed, someone else is in charge of payment. Semantic distinction, I know, but I think it also boosts clarity a touch.